// namor/dist/index.js
"use strict";
exports.__esModule = true;
exports.validate = exports.rawData = exports.valid_subdomain = exports.dictionaries = void 0;
var generate_1 = require("./generate");
var dict_1 = require("./dict");
exports.dictionaries = {
    reserved: (0, dict_1.getDictFile)("reserved.txt"),
    "default": (0, dict_1.getDict)("default"),
    rugged: (0, dict_1.getDict)("rugged")
};
function valid_subdomain(name, opts) {
    if (opts === void 0) { opts = {}; }
    var result = /^[\w](?:[\w-]{0,61}[\w])?$/.test(name);
    if (opts.reserved) {
        var strippedName = name.replace(/[^a-zA-Z0-9]/, "");
        var reservedWords = Array.isArray(opts.reserved)
            ? opts.reserved
            : exports.dictionaries.reserved;
        return result && reservedWords.indexOf(strippedName) === -1;
    }
    return result;
}
exports.valid_subdomain = valid_subdomain;
exports.generate = generate_1.generate;
exports.getDict = dict_1.getDict;
exports.getDictFile = dict_1.getDictFile;
exports.rawData = exports.dictionaries;
function validate(name, opts) {
    console.log("namor: `validate` has been renamed to `valid_subdomain`");
    return valid_subdomain(name, opts);
}
exports.validate = validate;
exports["default"] = {
    generate: generate_1.generate,
    getDict: dict_1.getDict,
    getDictFile: dict_1.getDictFile,
    valid_subdomain: valid_subdomain,
    validate: validate,
    rawData: exports.dictionaries
};
